api.ts

```diff
@@ -1423,23 +1440,46 @@ export function makeMetadataFormula(
   });
 }
 
-function makePropertyAutocompleteFormula(
-  execute: PropertyAutocompleteMetadataFunction,
-  options?: {connectionRequirement?: ConnectionRequirement},
-): PropertyAutocompleteMetadataFormula {
+/**
+ * Builds a formula to store in {@link SyncTableAutocompleters}.
+ *
+ * @hidden
+ */
+export function makePropertyAutocompleteFormula<SchemaT extends Schema>({
+  execute,
+  schema,
+  name,
+}: {
+  execute: PropertyAutocompleteMetadataFunction<Array<SchemaType<SchemaT>>>;
+  schema: SchemaT;
+  name: string;
+}): PropertyAutocompleteMetadataFormula<SchemaT> {
   if (!(execute instanceof Function)) {
-    throw new Error(`Value for propertyAutocomplete must be a function`);
+    throw new Error(`Value for execute must be a function`);
   }
-  return makeObjectFormula({
-    name: 'getPropertyAutocompleteMetadata',
-    description: 'Gets property autocomplete',
-    execute([], context) {
-      return execute(context as PropertyAutocompleteExecutionContext);
-    },
+
+  type ResultT = SchemaType<ArraySchema<SchemaT>>;
+
+  // The type SchemaType<ArraySchema<T>> is equivalent to Array<SchemaType<T>>, but typescript doesn't know
+  // that unless we do a cast.
+  const executeRetyped = execute as PropertyAutocompleteMetadataFunction<SchemaType<ArraySchema<SchemaT>>>;
+
+  const innerExecute = async ([]: ParamValues<[]>, context: ExecutionContext): Promise<ResultT> => {
```

Comment:
This is just because we have to bend this to something matching the shape of a regular `MetadataFormula`?